### PR TITLE
horace on fixed

### PR DESCRIPTION
**Pull request by abuts**
fix the horace_on script allowing to enable herbert, which correspond to another Horace

### CODE REVIEW

**Review by cmarooney-stfc — APPROVED** [review]

File: admin/horace_on.m.template [hunk 1]
```diff
@@ -30,27 +41,36 @@ if ~isempty(sw_start)
     spinw_on(default_spinw_path);
 end
 
+if nargin > 1
+    independent_herbert = true;
+else
+    independent_herbert = false;
+end
```
Comment (oerc0122) [inline admin/horace_on.m.template:48]: Why not 
```MATLAB
independent_herbert = exist('non_default_her_path', 'var') && ~isempty(non_default_her_path);
```
(or equivalent with if) which both makes intent clear and protects against later modification.

File: admin/horace_on.m.template [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
 %  safely switches Horace on
 %  horace_on()                         -- calls Horace with default settings
 %  horace_on(non_default_horace_path)  -- calls Horace with non-default Horace folder;
-%
+%  horace_on(non_default_horace_path,non_default_her_path)
+%                                      -- calls Horace with non-default Horace folder
+%                                         and non-default Herbert folder,
+%                                         not located within non-default
+%                                         Horace folder (for old versions
+%                                         of Horace/Herbert pairs)
```
Comment (oerc0122) [inline admin/horace_on.m.template:10]: The wording here is confusing
```suggestion
%                                      -- calls Horace with non-default Horace folder
%                                         and non-default Herbert folder,
%                                         for Herbert external to current
%                                         Horace folder (for old versions
%                                         of Horace/Herbert pairs). If empty
%                                         use default.
```

Comment (oerc0122) [reply admin/horace_on.m.template:10]: I would also consider changing `calls` to `loads`

File: admin/horace_on.m.template [hunk 3]
```diff
@@ -12,9 +17,15 @@ default_horace_path  = '${Horace_CORE}';
 % with horace_on.m script.
 default_spinw_path = '/usr/local/mprogs/spinW';
 %
-if exist('non_default_path','var') && (strcmpi(non_default_path,'where') || strcmpi(non_default_path,'which'))
-    path = find_default_hor_path(default_horace_path);
-    return;
+if exist('non_default_hor_path','var')
+    if (strcmpi(non_default_hor_path,'where') || strcmpi(non_default_hor_path,'which'))
```
Comment (oerc0122) [inline admin/horace_on.m.template:21]: Would also check for `isempty` here

File: admin/horace_on.m.template [hunk 4]
```diff
@@ -30,27 +41,36 @@ if ~isempty(sw_start)
...
+        init_herbert(herbert_path);
     catch ME
         disp('HORACE_ON:wrong_dependencies - cannot initiate Herbert');
         rethrow(ME);
     end
 else % reinitialize Herbert on where it is now.
-    her_path = fileparts(which('herbert_init.m'));
-    init_herbert(her_path);
+    if independent_herbert
+        init_herbert(non_default_her_path);
```
Comment (oerc0122) [inline admin/horace_on.m.template:65]: Why not move the `if independent_herbert` block outside the `if ~herbert_initiated` block and use `herbert_path` set there?

File: admin/horace_on.m.template [hunk 5]
```diff
@@ -70,7 +90,7 @@ function her_path = build_default_her_path(hor_path)
 if strcmp(hor_folder,'horace_core')
     her_path = fullfile(fp,'herbert_core');
 else
-    her_path = fullfile(hor_path,'herbert_core');    
+    her_path = fullfile(hor_path,'herbert_core');
```
Comment (oerc0122) [inline admin/horace_on.m.template:93]: If this is an external herbert, we don't want `her_path` defined from `hor_path`, it should be taken from where `herbert_init` is.

**Review by oerc0122 — CHANGES_REQUESTED** [review]

File: admin/horace_on.m.template (hunk 1 above)
Comment (abuts) [reply admin/horace_on.m.template:48]: I prefer avoiding checking for input variable existence, as in fact it only checks if the position of the variable is filled. 
I.e. if you have `my_fun(a,b)` and call it `my_fun(b)` the check inside `my_fun`  `exist('b','var')` would return `false`. So the check gives you false feeling that you are checking for variable existence when in fact you are checking for `nargin`. `isempty` can be added -- no harm.

Comment (oerc0122) [reply admin/horace_on.m.template:48]: This reasoning doesn't make sense to me. You are using positional arguments, so of course it only checks the position. Within the function the name of the outer variable doesn't mean anything. You are checking exactly what you asked for. If you want to use keyword arguments, then you can use keyword arguments within `inputParser` and subsequently check for presence.

```matlab
p = inputParser();
addRequired(p,'a');
addParameter(p,'b', []);
parse(p,a, b);

if p.Results.usingDefaults.b
   ...
end
```
But there are complaints that this is slow in performance critical functions (while I don't think this is).

Checking nargin adds a level of abstraction in determining which argument is present or not, reduces extensibility and maintainability for other programmers and in fact, with newer (2022a) MATLAB keyword syntax what you are checking is fundamentally wrong.
```matlab
my_fun(b='b')
% nargin -> 1
% b -> exists
```

File: admin/horace_on.m.template (hunk 4 above)
Comment (abuts) [reply admin/horace_on.m.template:65]: because `independent_herbert` is used in both Herbert branches.

File: admin/horace_on.m.template (hunk 5 above)
Comment (abuts) [reply admin/horace_on.m.template:93]: its exactly what the code doing. Its not building Herbert path from Horace if Herbert is independent

**Comment by cmarooney-stfc** [comment]
as this has been merged, moving to Done

**Pull request merged**